feat: interest rate

gnkz opened the pull request.

gnkz committed:
feat: debt interest rate

gnkz committed:
test: test interest claiming

gnkz committed:
feat: track earned interest

gnkz committed:
chore: rename variable

gnkz committed:
feat: add setter for max interest rate

coffee-the-dev commented on src/core/VaultManagerV6.sol:
```diff
@@ -61,17 +77,73 @@ contract VaultManagerV6 is IVaultManagerV5, UUPSUpgradeable, OwnableUpgradeable
         keroseneValuer = KeroseneValuer(_keroseneValuer);
+
+        uint256 interestIndex = INTEREST_PRECISION;
+        _interestIndexSnapshot = interestIndex;
```
won't `interestIndex` always be zero here?

gnkz replied:
no, the initial interest index is set to `INTEREST_PRECISION` and so its different from zero

coffee-the-dev commented on src/core/VaultManagerV6.sol:
```diff
@@ -61,17 +77,73 @@ contract VaultManagerV6 is IVaultManagerV5, UUPSUpgradeable, OwnableUpgradeable
+        uint256 totalNotes = dNft.totalSupply();
+        Dyad dyadCached = dyad;
+
+        for (uint256 i; i < totalNotes; i++) {
```
have you run a simulation to see how much gas this uses? I still think we should lazy initialize these as the users interact with the vault. For calculations, we can assume that if `mintedDyad > 0` and `interestIndex == 0` that `interestIndex = 1`

gnkz replied:
Yes, with the current amount of notes it uses ~24M gas which is around $550 at current gas and eth price. I replaced the loop with a lazy initialization tho this will increase the gas cost for all user interactions with the vault manager

coffee-the-dev commented on src/core/VaultManagerV6.sol:
```diff
@@ -61,17 +77,73 @@ contract VaultManagerV6 is IVaultManagerV5, UUPSUpgradeable, OwnableUpgradeable
+
+        if (newInterestRate < interestRate) {
+            interestRate = newInterestRate;
+        }
```
suggestion: emit an event here to indicate that the max interest rate has changed

gnkz replied:
done

coffee-the-dev commented on src/core/VaultManagerV6.sol:
```diff
@@ -61,17 +77,73 @@ contract VaultManagerV6 is IVaultManagerV5, UUPSUpgradeable, OwnableUpgradeable
         keroseneValuer = KeroseneValuer(_newKeroseneValuer);
     }
 
+    function setMaxInterestRate(uint256 _newMaxInterestRateBps) external onlyOwner {
```
this function is `setMaxInterestRate` but is setting the rate, not the max

gnkz replied:
my bad, fixed!

coffee-the-dev commented on src/core/VaultManagerV6.sol:
```diff
@@ -144,6 +224,12 @@ contract VaultManagerV6 is IVaultManagerV5, UUPSUpgradeable, OwnableUpgradeable
     /// @inheritdoc IVaultManagerV5
     function burnDyad(uint256 id, uint256 amount) public isValidDNft(id) {
         dyad.burn(id, msg.sender, amount);
+        uint256 currentNoteDebt = _accrueNoteInterest(id);
```
the dyad ERC20 contract itself is checking how much the note has minted when burning, so we need to actually mint the interest here before calling burn.

We should also special case uint256.max so it burns all debt (similar to other platforms) so users don't have to provide the exact amount to burn to clear all debt.

gnkz replied:
good catch. Added a function that will mint dyad if the amount that's getting repaid is greater than the amount that was minted. Also it clears all the debt if the amount is greater than the compounded debt

https://github.com/DyadStablecoin/contracts/pull/134/files#diff-644c15632cc794a8b8de95c3e7a1154b9d95a59da61acebea136a3f50e97f6cdR589

gnkz committed:
chore: fix set max interest rate

gnkz committed:
chore: resolve review comments

- emits event when max interest rate is updated
- removes loop to initialize interest rate system state
- fixes error that triggers underflow when burning dyad

gnkz committed:
fix: use correct variable
